Add start and end timestamp to task and play result in json callback

felixedel opened the pull request:
##### SUMMARY
Currently, the timestamp information is only provided directly by a few
Ansible modules (e.g. the command module, which shows the runtime of a
command per host result).
This change adds an 'overall' time information to all executed tasks. The
delta between both timestamps shows how long it took a task to finish
across all hosts/nodes.

Update: This information is now also available for plays.

This patch is also proposed for zuul and can be found here:
https://review.openstack.org/#/c/563888

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
lib/ansible/plugins/callback/json.py

##### ANSIBLE VERSION
```
2.6devel
```

felixedel committed:
Add start and end timestamp to task result in json callback

Currently, the timestamp information is only provided directly by a few
Ansible modules (e.g. the command module, which shows the runtime of a
command per host result).
This change adds an 'overall' time information to all executed tasks. The
delta between both timestamps shows how long it took a task to finish
across all hosts/nodes.

This patch is also proposed for zuul and can be found here:
https://review.openstack.org/#/c/563888

sivel commented:
I think I would prefer a couple of changes to this PR.

1. `start` and `end` should be nested under another key (maybe `duration`), so that it more clearly indicates what they are for.
1. This should not be limited to tasks, but extended to plays also.
1. For added clarity, please append a `Z` on the end of the output from isoformat, as any time representation not explicitly indicating a timezone offset is assumed to be local time.

    ```
    'start': '%sZ' % datetime.datetime.utcnow().isoformat()
    ```

felixedel committed:
Add missing timezone information to 'start' and 'end' timestamps

As the datetime.isoformat() function is missing the timezone information,
we assume it's local time.

felixedel committed:
Nest 'start' and 'end' timestamps in 'duration' field.

To clarify the purpose of those fields.

felixedel committed:
Add 'start' and 'end' timestamps also for plays

felix-schmidt commented:
Thanks for the feedback 😃 
I've adapted my change accordingly.

1. Fair point 😄 I've oriented myself on the result of the command module, which provides `start` and `end` as top-level fields, but afterwards I was also wondering if that is clear enough in this case.

2. Done. As I could not find an appropriate callback function like `v2_playbook_on_play_end()` or similar the end timestamp of a play is now also updated whenever a task finishes on a host. I've also moved the 'duration' field of a task inside the internal 'task' field to go in hand with the updated play result.

3. Done
